### PR TITLE
jira/NMS-13098

### PR DESCRIPTION
**Pull request by jadyhou**
Skip process Action while there is no Action or an Action with empty "Statement" or content.

### All Contributors

* [ ] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ ] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [ ] Have you made a comment in that issue which points back to this PR?
* [ ] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/${JIRA-ISSUE-NUMBER}

### CODE REVIEW

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Review by RangerRick — APPROVED** [review]

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java
Comment (mattixtech) [inline opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java:100]: Is `null` a legal state for `m_statement`? The XML annotation implies it is required. Under what circumstances is it actually legal for it to be null?

If it is illegal for it to be null we could update the getter to assert non-null with an `IllegalStateException` or ideally post-deserialization do that to guarantee that invariant is checked.

The fact that an NPE is blowing up deeper in the stack might just be an indication that we created an illegal object earlier on.

---

If it is in fact legal to be `null` we can explicitly indicate that to our callers by refactoring the getter to return an Optional. We can then get rid of this `hasStatement` method.

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [inline opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: This is a bit suspect, is it really legal for something called an `ActionProcessor` to have a `null` action? Under what circumstance is that allowed?

I realize it has a method to check if it has an action, but it looks like you are the first user of that and the fact it was present smells a little hacky to begin with.

**Review by mattixtech — CHANGES_REQUESTED** [review]
I have a few general questions. I want to make sure we aren't just masking the real issue by adding these null checks.

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java
Comment (mattixtech) [reply opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java:100]: Forgot to mention, the setter also implies that null is illegal. So pretty sure this shouldn't be allowed null.

Comment (jadyhou) [reply opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java:100]: Thanks Matthew. That's a very good point. The statement could not be null, it would throw exception by xml read before reaching the line of the code.

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: As far as I can find out, the reason ActionProcessor can have a null action is when the user input an "action name" that isn't in the action list, which could be caused by a typo by user.

Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: I think at the point, we might need to decide how to handle this senario. Do we just skip it or do we at least log this so customer know there is something wrong or do we really want to throw an exception?

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java
Comment (mattixtech) [reply opennms-config-model/src/main/java/org/opennms/netmgt/config/vacuumd/Action.java:100]: Ok, if we are confident that `null` is an illegal state for this type and that we already guard against that invariant let's remove the new redundant code that checks that.

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: Ok so as far as I can tell, an `AutomationProcessor` delegates to an `ActionProcessor`. We added null checking to the `ActionProcessor` to guard against a null `Action` which probably doesn't make sense.

I'm tempted to say we should instead move the check up a level.

We could say that it is illegal for an `ActionProcessor` to have a `null` `Action`. Then we can guard against that via asserting an `IllegalArgumentException` in the `ActionProcessor` class.

Now in the constructor for `AutomationProcessor`, rather than set it's `ActionProcessor` to one with a null `Action`, we set the `ActionProcessor` instance to null itself.

Now you can add new guards in the processing logic to see if you have a non-null `ActionProcessor` or not and take action accordingly. Could also update the getter to return an `Optional` also (although the getter doesn't seem to be used anymore).

Does this make sense?

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: Should be able to back out most of the changes made and also those in https://github.com/OpenNMS/opennms/pull/3285 if we follow through with this.

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: Sure, just one second.

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:710]: Additionally, if we think it is an error case we could blow up the constructor of `AutomationProcessor` and log a warning rather than set its delegate to `null` and guard against it. I'm not familiar enough with the problem to know which is the right solution.

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: I don't like null, and don't like assert. So I pick the second option and log the warning : -).

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: What if we propagate the error up another level to the `AutomationProcessor` constructor. Can we just blow up there if it finds a null? Then we don't need any of this guard logic anymore.

Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: Either way would work. I think these are just different coding styles. the reason I don't like to put the null check in  AutomationProcessor constructor is that I think it doesn't hurt to have an AutomationProcessor with an "Action" that has an empty field "action". I just don't like to blow up the whole thing just because this null "action" doesn't make sense. Also throwing exceptions may impact performance, and this is automation keeps on running. Also, I want to log the warning to let the user know what's wrong and how they can fix it, put the log too deep in the code make the warning seems not too related to that place.   
Of course, there will also be a bright side to handling the error in AutomationProcessor constructor.  
Either way works for me. If we feel put the checking inside constructor more fits OpenNMS style, I have no problem doing that.

Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: By the way, I'm not sure what's our plan for Vacuumd, I guess we definitely will need to redesign it to make it user-friendly but not letting the user write SQL :P. Also configuring via XML is not desirable :D

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: In this case I don't think it's just a style issue. I think the question is whether an `AutomationProcessor` is even valid in the case where we couldn't find an `Action` for it. I suspect the answer is now. In that case we want to fail as early as possible so we can accurately describe what went wrong and where.

Here is a patch for what I had in mind: https://github.com/OpenNMS/opennms/compare/mbrooks/null_action

Please take a look at my proposed change and let me know what you think.

This "fixes" the NPE by turning it into a proper error. I think the other attempts we were trying here to fix it were addressing the problem at too low a level and didn't really address the underlying problem.

With that out of the way, I think there is a separate issue here since an `Action`'s `Statement` can have empty content, does that make sense? Calling `getActionSQL()` can return an empty String which is a bit odd, it could probably be changed to an `Optional<String>` instead. Might not be worth addressing in this PR though as I think it is a separate issue.

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: Please note I am making an assumption here about how an `AutomationProcessor` works and what its invariants are based on little understanding of the domain here so correct me if I'm wrong anyone else on this review.

Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: Happy Monday Matthew! The patch you share works for me too. Just one note, we probably don't need to check if the "actionname" is null, because the application will fail to start if we leave "actionname" blank.  I have another idea, how about we throw the exception  in where the null come from.  I'm attaching it here in the comment.
    public synchronized Action getAction(String actionName) {
        for (Action act : getActions()) {
            if (act.getName().equals(actionName)) {
                return act;
            }
        }
//        return null;
        throw new IllegalArgumentException("Could not find an action for automation action named '" + actionName + "'");
    }

If the requirement is user must have the "action" and it has to be valid, then this is the only code change we need to make.

Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:709]: Hi Jane, good idea addressing the issue there, however I think we could improve it even a bit further. Rather than throw an exception in `getAction` we can simply return a `Optional<Action>` and leave the exception throwing up to the caller still. This allows the API of `getAction` to remain flexible.

So our `AutomationProcessor` constructor would look like this:
```
        String actionName = automation.getActionName();
        Action actionForAutomation = VacuumdConfigFactory.getInstance()
                .getAction(actionName)
                .orElseThrow(() -> new IllegalArgumentException("Could not find an action for automation action named '" + actionName + "'"));
```

I think with this solution we will be good to go.

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (jadyhou) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java]: Hi Matthew. I like this idea! I've updated the code and tested on my machine.

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [inline opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:889]: We can delete this now right?

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [inline opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:712]: Could you please revert the whitespace/formatting changes on lines 712, 715, 726. Thanks

**Review by mattixtech — CHANGES_REQUESTED** [review]
Looking good! Just a couple minor comments.

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [reply opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java]: Thanks Jane, looks good.

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Review by mattixtech — APPROVED** [review]
LGTM 🚀

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java
Comment (mattixtech) [inline opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java:708]: Looks like there is still an extra line here to get rid of.

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome jane.hou (jane.hou@xyleminc.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username jane.hou (jane.hou@xyleminc.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by mattixtech** [comment]
@jadyhou Hi Jane, For PRs like this can you make sure to "squash merge" them going forward when you merge them to `develop`. It is easier to follow the changes if they are all grouped together in one commit. The exception to this rule would be if you have a few commits where there is a good reason to keep them discrete for organization/readability/ease of rollback reasons. But if the commit text is something like "clean up" then that probably doesn't qualify 😛 . Thanks!

**Comment by jadyhou** [comment]
@mattixtech Sure Matthew, that is a very reasonable request, I'd prefer to do it that way too. Let me figure out how to do squash merge :).

**Comment by indigo423** [comment]
@jadyhou I think we should at least wait for CircleCI to pass as the last barrier before we click the merge button :)

Screenshot 2021-02-09 at 13 40 38

**Comment by jadyhou** [comment]
@indigo423 Yeah, this is annoying to me too. I did sign the agreement. And this email address is not in my Github account profile.

**Comment by indigo423** [comment]
@jadyhou I've seen the same error when I try to run our latest develop build (bleeding) from here: https://49733-9377198-gh.circle-artifacts.com/0/test-artifacts/logs/org.opennms.smoketest.MenuHeaderIT-null/opennms/container_stdout_stderr

The only way to get it Horizon starting is by disabling Vaccumd in the service-configuration. Do you have some ideas?

**Comment by jadyhou** [comment]
@indigo423 Yes, I'll revert the merge first.

**Comment by indigo423** [comment]
> @indigo423 Yeah, this is annoying to me too. I did sign the agreement. And this email address is not in my Github account profile.

Ok, what I've seen is the git commit was made with your old mail address which didn't match the OCA as you have mentioned. Our bot checks if your git config mail addresses match the OCA list and stumble over the four commits with your old mail address. With new commits and PR's it won't happen. We probably need to change the commits from these 4 commits:

```
commit 93a508ec982850f5fcef7ec56dac8275a7770706
Merge: 45c09df3afa 54120cd5fff
Author: jane.hou <jane.hou@xyleminc.com>
Date:   Thu Feb 4 09:31:52 2021 -0500

    Merge remote-tracking branch 'origin/jira/NMS-13098' into jira/NMS-13098

commit 45c09df3afa15b40c5c45ef4ffecf155c59e2c03
Author: jane.hou <jane.hou@xyleminc.com>
Date:   Thu Feb 4 09:31:20 2021 -0500

    NMS-13098. Skip process Action while there is no Action configured in Vacuumd

commit 2355d5518701974fb12696a7c3958b8660cda285
Author: jane.hou <jane.hou@xyleminc.com>
Date:   Tue Feb 2 09:49:31 2021 -0500

    NMS-13098. Remove some Intellij warning.

commit 98d414965eb9c8024cc5ea4548cf7d8dbac32d6c
Author: jane.hou <jane.hou@xyleminc.com>
Date:   Tue Feb 2 09:35:38 2021 -0500

    NMS-13098. Fix NPE.
```
Maybe @RangerRick can help us here with his wonderful git Kung Fu :)

**Comment by jadyhou** [comment]
I'll have another pull request to log the error then the CI should be fine.

**Pull request merged**
